### PR TITLE
Implementation of project groupings

### PR DESCRIPTION
**Pull request by tulinkry**
I'm getting closer to #480, which I suppose is a prerequisite of the authorization layer.

Groups are taken from configuration file (example below) which is given to the jar run (-W option) and it is therefore propagated to the web application through the socket listener (then it lives in the main configuration file). Project can be in multiple groups.

A new class ProjectHelper is introduced to wrap operations around env.projects so that it can be easily changed later to support authorizator filtering of projects or groups. 

There are couple of screenshots:

1) Groups as a list on the main page. User can click on them and their content gets expanded.
screen
2) UI change of the selectbox to support top level (HTML restriction) groups
screen1
3) expanded groups
screen2

Working demo available [here](http://virtual99.cz.oracle.com/groups/).

Tested in IE 9, Firefox and Chrome.

Example of configuration:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<java version="1.8.0_65" class="java.beans.XMLDecoder">
 <object class="org.opensolaris.opengrok.configuration.Configuration" id="Configuration0">
	<void method="addGroup">
		<object class="org.opensolaris.opengrok.configuration.Group">
			<void property="name">
				<string>Haxtip</string>
			</void>
			<void property="pattern">
				<string>rfe.*|sql</string>
			</void>
			<void method="addGroup">
	 			<object class="org.opensolaris.opengrok.configuration.Group">
	 				<void property="name">
	 					<string>Git</string>
	 				</void>
	 				<void property="pattern">
	 					<string>haxtip|Replicator</string>
	 				</void>
			 		<void method="addGroup">
			 			<object class="org.opensolaris.opengrok.configuration.Group">
			 				<void property="name">
			 					<string>OpenGrok</string>
			 				</void>
			 				<void property="pattern">
			 					<string>metric|git|OpenGrok</string>
			 				</void>
			 			</object>
			 		</void>
	 			</object>
			</void>
		</object>
	</void>
	<void method="addGroup">
		<object class="org.opensolaris.opengrok.configuration.Group">
			<void property="name">
				<string>Python</string>
			</void>
			<void property="pattern">
				<string>Downloader</string>
			</void>
		</object>
	</void>
 </object>
</java>
```

### CODE REVIEW

File: src/org/opensolaris/opengrok/configuration/Project.java [hunk 1]
```diff
@@ -26,11 +26,20 @@
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.util.Comparator;
 
 /**
  * Placeholder for the information that builds up a project
  */
 public class Project {
+
```
Comment (tarzanek) [inline src/org/opensolaris/opengrok/configuration/Project.java]: if you properly implement equals and hashCode, there is no need for comparator ...

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 2]
```diff
@@ -827,6 +838,25 @@ private static Configuration decodeObject(InputStream in) throws IOException {
         if (!(ret instanceof Configuration)) {
             throw new IOException("Not a valid config file");
         }
+        
+        for (Iterator<Group> it = ((Configuration)ret).getGroups().iterator(); it.hasNext();) {
```
Comment (tarzanek) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: for (Group g:getgroups() ) ...

if you use Netbeans, please notice the left side yellow/orange exclamation marks - hints - usually it should make you aware

Also this check is overengineering(not sure how did you come to such overly complex cycle in cycle ;) ... KISS!) - if you need a deduplicated list, use appropriate structure and make sure your object properly does hashCode and equals implementation

File: src/org/opensolaris/opengrok/configuration/Group.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+     * names along all group's children's siblings.
+     *
+     * Same names do not violate this condition if they are in different
+     * branches. Same names do not violate this condition if they are in the
+     * same branch but they are not siblings
+     *
+     * @param g Group
+     * @return true if it satisfies conditions above
+     */
+    public static boolean isValid(Group g) {
```
Comment (tarzanek) [inline src/org/opensolaris/opengrok/configuration/Group.java]: I don't get what do we try to solve with this :)
But it also seems too complicated for me and that doesn't feel OK

**Comment by tarzanek** [comment]
looks good, but I have some concerns about the design from code point of view + there is some design/restrictions, that slip my understanding.
Let's meet in the office and talk this out (since we have the luxury of meeting in person, I will skip comments here to save our time). Wed/Thu , check my internal calendar when I am free, I will jump in your office, ev. schedule a meeting and invite me & @vladak pls

**Comment by tulinkry** [comment]
I modified the solution with native Java structures, demo isn't available now.

**Comment by tulinkry** [comment]
I removed that unused import.

"I have a bad feeling document.ready is already called somewhere, but the old code might have been refactored, can you double verify no other jsp 9mast,head, ... list, ...) is calling .ready ?
Also I am not sure if jquery permits several .ready calls"

- ready() is already called twice (table sorter and the inteligence window init)
- ready() is a callback register, so it's not problem when we register multiple handlers for ready event - each of them is executed

There was a little problem when user defined no group. I fixed that so now it's compatible with both modes - using groups or without groups.

**Comment by tarzanek** [comment]
merging now, so we have time to stabilize it

**Comment by tarzanek** [comment]
thank you Krystof!
(keep them coming! :) )

**Pull request merged**